lib/sigstore/signer.rb

```diff
@@ -0,0 +1,246 @@
+      csr = OpenSSL::X509::Request.new
+
+      csr.version = 0 # TODO: check
+      # TODO: proper name
+      # The subject in the CertificationRequestInfo is an X.501 RelativeDistinguishedName.
+      # The value of the RelativeDistinguishedName SHOULD be the subject of the authentication token;
+      # its type MUST be the type identified in the Fulcio instance’s public configuration.
+      csr.subject = OpenSSL::X509::Name.new [%w[CN noone], %w[DC example]]
+      csr.public_key = keypair
+      # TODO: digest from trusted root config
+      csr.sign keypair, "SHA256"
+
+      {
+        credentials: {
+          oidc_identity_token: @jwt
+        },
+        certificate_signing_request: Internal::Util.base64_encode(csr.to_pem)
+      }
+    end
+
+    def fetch_cert(csr)
+      uri = URI.parse @trusted_root.certificate_authority_for_signing.uri
+      uri = URI.join(uri, "api/v2/signingCert")
+      resp = Net::HTTP.post(
+        uri,
+        JSON.dump(csr),
+        { "Content-Type" => "application/json" }
+      )
+
+      resp.value
+      resp_body = JSON.parse(resp.body)
+
+      chain = resp_body.fetch("signedCertificateEmbeddedSct").fetch("chain")
+                       .fetch("certificates").map { |pem| Internal::X509::Certificate.read(pem) }
+      logger.debug { "Fetched cert from fulcio" }
+      chain
+    end
+
+    def verify_chain(chain)
+      # Perform certification path validation (RFC 5280 §6) of the returned certificate chain with the pre-distributed
+      # Fulcio root certificate(s) as a trust anchor.
+
+      x509_store = OpenSSL::X509::Store.new
+      chain = chain.map(&:openssl)
+      leaf = chain.shift
+      root = chain.pop # TODO: use root from trusted root as trust anchor
+      x509_store.add_cert root
+      raise x509_store.error_string unless x509_store.verify(leaf, chain)
+
+      logger.debug { "verified chain" }
+
+      # TODO: verify SCT in leaf
+      # Extract a SignedCertificateTimestamp, which may be embedded as an X.509 extension in the leaf certificate or
+      # attached separately in the SigningCertificate returned from the Identity Service.
+      # Verify this SignedCertificateTimestamp as in RFC 9162 §8.1.3, using the root certificate from
+      # the Certificate Transparency Log.
+
+      # TODO: verify leaf subject
+      # Check that the leaf certificate contains the subject from the certificate signing request and encodes the
+      # appropriate AuthenticationServiceIdentifier in an extension with OID 1.3.6.1.4.1.57264.1.8.
+
+      [leaf, x509_store.chain]
+    end
+
+    def sign_payload(payload, key)
+      # TODO: derive correct digest from what the registry supports
+      # The Signer MAY pre-hash the payload using a hash algorithm from the registry (Spec: Sigstore Registries) for
+      # compatibility with some signing metadata formats (see §Submission of Signing Metadata to Transparency Service).
+      key.sign("SHA256", payload)
+    end
+
+    # TODO: implement
```

Comment:
- [ ] implement `submit_signature_hash_to_timstamping_service`

Reply:
... in practice, is it OK to skip this for now, since all clients seem to allow using the rekor entry as a trusted source of timestamping?

Reply:
Yep, in practice all clients use the inclusion promise or signed checkpoint from Rekor as a source of signed time.